# BÃ¶lÃ¼m 1: YÃ¶nteme genel bakÄ±ÅŸ ve manuel test

<span class="ai-translation-notice">:material-information-outline:{ .ai-translation-notice-icon } Yapay Zeka Destekli Ã‡eviri - [daha fazla bilgi ve iyileÅŸtirme Ã¶nerileri](https://github.com/nextflow-io/training/blob/master/TRANSLATING.md)</span>

Toplu RNAseq verilerini iÅŸlemek ve analiz etmek iÃ§in birden fazla geÃ§erli yÃ¶ntem bulunmaktadÄ±r.
Bu kurs iÃ§in, [Babraham EnstitÃ¼sÃ¼](https://www.babraham.ac.uk/)'nden Dr. Simon Andrews ve Laura Biggins tarafÄ±ndan [burada](https://www.bioinformatics.babraham.ac.uk/training/RNASeq_Course/Analysing%20RNA-Seq%20data%20Exercise.pdf) aÃ§Ä±klanan yÃ¶ntemi takip ediyoruz.

AmacÄ±mÄ±z, aÅŸaÄŸÄ±daki iÅŸleme adÄ±mlarÄ±nÄ± uygulayan bir iÅŸ akÄ±ÅŸÄ± geliÅŸtirmektir: toplu RNAseq Ã¶rneÄŸindeki okumalar Ã¼zerinde ilk kalite kontrolÃ¼nÃ¼ Ã§alÄ±ÅŸtÄ±rmak, okumalardan adaptÃ¶r dizilerini kÄ±rpmak, okumalarÄ± referans genoma hizalamak ve kapsamlÄ± bir kalite kontrol (QC) raporu oluÅŸturmak.

<figure class="excalidraw">
--8<-- "docs/en/docs/nf4_science/rnaseq/img/preprocess.svg"
</figure>

- **FASTQC:** KÄ±rpma Ã¶ncesi okuma verileri Ã¼zerinde FastQC kullanarak QC gerÃ§ekleÅŸtirin
- **TRIM_GALORE:** Trim Galore kullanarak adaptÃ¶r dizilerini kÄ±rpÄ±n ve kÄ±rpma sonrasÄ± QC gerÃ§ekleÅŸtirin (Cutadapt ve FastQC'yi bir araya getirir)
- **HISAT2_ALIGN:** Hisat2 kullanarak okumalarÄ± referans genoma hizalayÄ±n
- **MULTIQC:** MultiQC kullanarak kapsamlÄ± bir QC raporu oluÅŸturun

Ancak, herhangi bir iÅŸ akÄ±ÅŸÄ± kodu yazmaya baÅŸlamadan Ã¶nce, komutlarÄ± bazÄ± test verileri Ã¼zerinde manuel olarak deneyeceÄŸiz.
Ä°htiyacÄ±mÄ±z olan araÃ§lar GitHub Codespaces ortamÄ±nda yÃ¼klÃ¼ olmadÄ±ÄŸÄ±ndan, bunlarÄ± konteynerler aracÄ±lÄ±ÄŸÄ±yla kullanacaÄŸÄ±z (bkz. [Merhaba Konteynerler](../../hello_nextflow/05_hello_containers.md)).

!!! note "Not"

     `nf4-science/rnaseq` dizininde olduÄŸunuzdan emin olun. `pwd` yazdÄ±ÄŸÄ±nÄ±zda gÃ¶sterilen yolun son kÄ±smÄ± `rnaseq` olmalÄ±dÄ±r.

---

## 1. Ä°lk QC ve adaptÃ¶r kÄ±rpma

Hem `fastqc` hem de `trim_galore` yÃ¼klÃ¼ bir konteyner imajÄ±nÄ± Ã§ekeceÄŸiz, etkileÅŸimli olarak baÅŸlatacaÄŸÄ±z ve Ã¶rnek veri dosyalarÄ±ndan biri Ã¼zerinde kÄ±rpma ve QC komutlarÄ±nÄ± Ã§alÄ±ÅŸtÄ±racaÄŸÄ±z.

### 1.1. Konteyneri Ã§ekin

```bash
docker pull community.wave.seqera.io/library/trim-galore:0.6.10--1bf8ca4e1967cd18
```

Bu, sistem imajÄ± indirirken aÅŸaÄŸÄ±daki konsol Ã§Ä±ktÄ±sÄ±nÄ± verir:

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    0.6.10--1bf8ca4e1967cd18: Pulling from library/trim-galore
    dafa2b0c44d2: Pull complete
    dec6b097362e: Pull complete
    f88da01cff0b: Pull complete
    4f4fb700ef54: Pull complete
    92dc97a3ef36: Pull complete
    403f74b0f85e: Pull complete
    10b8c00c10a5: Pull complete
    17dc7ea432cc: Pull complete
    bb36d6c3110d: Pull complete
    0ea1a16bbe82: Pull complete
    030a47592a0a: Pull complete
    32ec762be2d0: Pull complete
    d2cb90387285: Pull complete
    Digest: sha256:4f00e7b2a09f3c8d8a9ce955120e177152fb1e56f63a2a6e186088b1250d9907
    Status: Downloaded newer image for community.wave.seqera.io/library/trim-galore:0.6.10--1bf8ca4e1967cd18
    community.wave.seqera.io/library/trim-galore:0.6.10--1bf8ca4e1967cd18
    ```

### 1.2. Konteyneri etkileÅŸimli olarak baÅŸlatÄ±n

```bash
docker run -it -v ./data:/data community.wave.seqera.io/library/trim-galore:0.6.10--1bf8ca4e1967cd18
```

<!--
??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console

    ```
-->

Ä°steminiz `(base) root@b645838b3314:/tmp#` gibi bir ÅŸeye dÃ¶nÃ¼ÅŸecek, bu da artÄ±k konteynerin iÃ§inde olduÄŸunuzu gÃ¶sterir.

Komutun `-v ./data:/data` kÄ±smÄ±, `data/` dizininin iÃ§eriÄŸine konteyner iÃ§inden eriÅŸmemizi saÄŸlayacaktÄ±r.

```bash
ls /data/reads
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    ENCSR000COQ1_1.fastq.gz  ENCSR000COQ2_2.fastq.gz  ENCSR000COR2_1.fastq.gz  ENCSR000CPO1_2.fastq.gz
    ENCSR000COQ1_2.fastq.gz  ENCSR000COR1_1.fastq.gz  ENCSR000COR2_2.fastq.gz  ENCSR000CPO2_1.fastq.gz
    ENCSR000COQ2_1.fastq.gz  ENCSR000COR1_2.fastq.gz  ENCSR000CPO1_1.fastq.gz  ENCSR000CPO2_2.fastq.gzO
    ```

### 1.3. Ä°lk `fastqc` komutunu Ã§alÄ±ÅŸtÄ±rÄ±n

Okuma verileri Ã¼zerinde kalite kontrol metriklerini toplamak iÃ§in `fastqc`'yi Ã§alÄ±ÅŸtÄ±ralÄ±m.

```bash
fastqc /data/reads/ENCSR000COQ1_1.fastq.gz
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    application/gzip
    Started analysis of ENCSR000COQ1_1.fastq.gz
    Approx 5% complete for ENCSR000COQ1_1.fastq.gz
    Approx 10% complete for ENCSR000COQ1_1.fastq.gz
    Approx 15% complete for ENCSR000COQ1_1.fastq.gz
    Approx 20% complete for ENCSR000COQ1_1.fastq.gz
    Approx 25% complete for ENCSR000COQ1_1.fastq.gz
    Approx 30% complete for ENCSR000COQ1_1.fastq.gz
    Approx 35% complete for ENCSR000COQ1_1.fastq.gz
    Approx 40% complete for ENCSR000COQ1_1.fastq.gz
    Approx 45% complete for ENCSR000COQ1_1.fastq.gz
    Approx 50% complete for ENCSR000COQ1_1.fastq.gz
    Approx 55% complete for ENCSR000COQ1_1.fastq.gz
    Approx 60% complete for ENCSR000COQ1_1.fastq.gz
    Approx 65% complete for ENCSR000COQ1_1.fastq.gz
    Approx 70% complete for ENCSR000COQ1_1.fastq.gz
    Approx 75% complete for ENCSR000COQ1_1.fastq.gz
    Approx 80% complete for ENCSR000COQ1_1.fastq.gz
    Approx 85% complete for ENCSR000COQ1_1.fastq.gz
    Approx 90% complete for ENCSR000COQ1_1.fastq.gz
    Approx 95% complete for ENCSR000COQ1_1.fastq.gz
    Analysis complete for ENCSR000COQ1_1.fastq.gz
    ```

Bu Ã§ok hÄ±zlÄ± Ã§alÄ±ÅŸmalÄ±dÄ±r.
Ã‡Ä±ktÄ± dosyalarÄ±nÄ± orijinal verilerle aynÄ± dizinde bulabilirsiniz:

```bash
ls /data/reads/ENCSR000COQ1_1_fastqc*
```

<!-- switch to tree -->

```console title="Ã‡Ä±ktÄ±"
/data/reads/ENCSR000COQ1_1_fastqc.html  /data/reads/ENCSR000COQ1_1_fastqc.zip
```

### 1.4. `trim_galore` ile adaptÃ¶r dizilerini kÄ±rpÄ±n

Åimdi adaptÃ¶r dizilerini kÄ±rpmak ve kÄ±rpma sonrasÄ± QC metriklerini toplamak iÃ§in Cutadapt ve FastQC'yi bir araya getiren `trim_galore`'Ä± Ã§alÄ±ÅŸtÄ±ralÄ±m.

```bash
trim_galore --fastqc /data/reads/ENCSR000COQ1_1.fastq.gz
```

`--fastqc` bayraÄŸÄ±, kÄ±rpma tamamlandÄ±ktan sonra komutun otomatik olarak bir QC toplama adÄ±mÄ± Ã§alÄ±ÅŸtÄ±rmasÄ±nÄ± saÄŸlar.

_Ã‡Ä±ktÄ± Ã§ok ayrÄ±ntÄ±lÄ± olduÄŸu iÃ§in aÅŸaÄŸÄ±daki kÄ±saltÄ±lmÄ±ÅŸtÄ±r._

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    Multicore support not enabled. Proceeding with single-core trimming.
    Path to Cutadapt set as: 'cutadapt' (default)
    Cutadapt seems to be working fine (tested command 'cutadapt --version')
    Cutadapt version: 4.9
    single-core operation.
    igzip command line interface 2.31.0
    igzip detected. Using igzip for decompressing

    <...>

    Analysis complete for ENCSR000COQ1_1_trimmed.fq.gz
    ```

Ã‡Ä±ktÄ± dosyalarÄ±nÄ± Ã§alÄ±ÅŸma dizininde bulabilirsiniz:

```bash
ls ENCSR000COQ1_1*
```

```console title="Ã‡Ä±ktÄ±"
ENCSR000COQ1_1.fastq.gz_trimming_report.txt  ENCSR000COQ1_1_trimmed_fastqc.html
ENCSR000COQ1_1_trimmed.fq.gz                 ENCSR000COQ1_1_trimmed_fastqc.zip
```

### 1.5. Ã‡Ä±ktÄ± dosyalarÄ±nÄ± konteyner dÄ±ÅŸÄ±ndaki dosya sistemine taÅŸÄ±yÄ±n

Konteyner iÃ§inde kalan her ÅŸey gelecekteki Ã§alÄ±ÅŸmalara eriÅŸilemez olacaÄŸÄ±ndan bunlarÄ± yeni bir dizine taÅŸÄ±yalÄ±m.

```bash
mkdir /data/trimmed
mv ENCSR000COQ1_1* /data/trimmed
```

### 1.6. Konteynerden Ã§Ä±kÄ±n

```bash
exit
```

---

## 2. OkumalarÄ± referans genoma hizalayÄ±n

`hisat2`'nin yÃ¼klÃ¼ olduÄŸu bir konteyner imajÄ±nÄ± Ã§ekeceÄŸiz, etkileÅŸimli olarak baÅŸlatacaÄŸÄ±z ve RNAseq verilerini referans genoma hizalamak iÃ§in hizalama komutunu Ã§alÄ±ÅŸtÄ±racaÄŸÄ±z.

### 2.1. `hisat2` konteynerini Ã§ekin

```bash
docker pull community.wave.seqera.io/library/hisat2_samtools:5e49f68a37dc010e
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    Unable to find image 'community.wave.seqera.io/library/hisat2_samtools:5e49f68a37dc010e' locally
    5e49f68a37dc010e: Pulling from library/hisat2_samtools
    dafa2b0c44d2: Already exists
    dec6b097362e: Already exists
    f88da01cff0b: Already exists
    4f4fb700ef54: Already exists
    92dc97a3ef36: Already exists
    403f74b0f85e: Already exists
    10b8c00c10a5: Already exists
    17dc7ea432cc: Already exists
    bb36d6c3110d: Already exists
    0ea1a16bbe82: Already exists
    030a47592a0a: Already exists
    e74ed5dd390b: Pull complete
    abfcf0185e51: Pull complete
    Digest: sha256:29d8e1a3172a2bdde7be813f7ebec22d331388194a7c0de872b4ccca4bed8f45
    Status: Downloaded newer image for community.wave.seqera.io/library/hisat2_samtools:5e49f68a37dc010e
    ```

### 2.2. `hisat2` konteynerini etkileÅŸimli olarak baÅŸlatÄ±n

```bash
docker run -it -v ./data:/data community.wave.seqera.io/library/hisat2_samtools:5e49f68a37dc010e
```

Komut daha Ã¶ncekiyle aynÄ±dÄ±r, ilgili konteyner URI'si deÄŸiÅŸtirilmiÅŸtir.

### 2.3. Hisat2 genom dizin dosyalarÄ±nÄ± oluÅŸturun

Hisat2, genom referansÄ±nÄ±n Ã§ok Ã¶zel bir biÃ§imde saÄŸlanmasÄ±nÄ± gerektirir ve saÄŸladÄ±ÄŸÄ±mÄ±z `genome.fa` FASTA dosyasÄ±nÄ± doÄŸrudan kullanamaz, bu nedenle bu fÄ±rsatÄ± ilgili kaynaklarÄ± oluÅŸturmak iÃ§in kullanacaÄŸÄ±z.

```bash
hisat2-build /data/genome.fa genome_index
```

Ã‡Ä±ktÄ± Ã§ok ayrÄ±ntÄ±lÄ± olduÄŸundan aÅŸaÄŸÄ±daki kÄ±saltÄ±lmÄ±ÅŸtÄ±r:

<!-- TODO: switch to full output -->

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    Settings:
      Output files: "genome_index.*.ht2"
    <...>
    Total time for call to driver() for forward index: 00:00:16
    ```

Bu, Ã§alÄ±ÅŸma dizininde bulabileceÄŸiniz birden fazla genom dizin dosyasÄ± oluÅŸturur.

```bash
ls genome_index.*
```

```console title="Ã‡Ä±ktÄ±"
genome_index.1.ht2  genome_index.3.ht2  genome_index.5.ht2  genome_index.7.ht2
genome_index.2.ht2  genome_index.4.ht2  genome_index.6.ht2  genome_index.8.ht2
```

BunlarÄ± birazdan kullanacaÄŸÄ±z, ancak Ã¶nce bu genom dizin dosyalarÄ±yla sÄ±kÄ±ÅŸtÄ±rÄ±lmÄ±ÅŸ bir tarball oluÅŸturalÄ±m; bunlara daha sonra ihtiyacÄ±mÄ±z olacak ve bunlarÄ± oluÅŸturmak tipik olarak bir iÅŸ akÄ±ÅŸÄ±nÄ±n parÃ§asÄ± olarak yapmak istediÄŸimiz bir ÅŸey deÄŸildir.

```bash
tar -czvf /data/genome_index.tar.gz genome_index.*
```

Bu, genom dizin dosyalarÄ±nÄ± iÃ§eren bir `genome_index.tar.gz` tarball'Ä±nÄ± dosya sistemimizde `data/` dizininde saklar, bu da bu kursun 2. BÃ¶lÃ¼mÃ¼nde iÅŸimize yarayacaktÄ±r.

### 2.4. `hisat2` komutunu Ã§alÄ±ÅŸtÄ±rÄ±n

Åimdi hizalama komutunu Ã§alÄ±ÅŸtÄ±rabiliriz, bu komut `hisat2` ile hizalama adÄ±mÄ±nÄ± gerÃ§ekleÅŸtirir ve ardÄ±ndan Ã§Ä±ktÄ±yÄ± bir BAM dosyasÄ± olarak yazmak iÃ§in `samtools`'a aktarÄ±r.

Okuma verisi girdisi, Ã¶nceki adÄ±mda `trim_galore` ile oluÅŸturduÄŸumuz `/data/trimmed/ENCSR000COQ1_1_trimmed.fq.gz` dosyasÄ±dÄ±r.

```bash
hisat2 -x genome_index -U /data/trimmed/ENCSR000COQ1_1_trimmed.fq.gz \
    --new-summary --summary-file ENCSR000COQ1_1_trimmed.hisat2.log | \
    samtools view -bS -o ENCSR000COQ1_1_trimmed.bam
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    HISAT2 summary stats:
            Total reads: 27816
                    Aligned 0 time: 1550 (5.57%)
                    Aligned 1 time: 25410 (91.35%)
                    Aligned >1 times: 856 (3.08%)
            Overall alignment rate: 94.43%
    ```

Bu, Ã§ok kÃ¼Ã§Ã¼k bir test dosyasÄ± olduÄŸu iÃ§in neredeyse anÄ±nda Ã§alÄ±ÅŸÄ±r.
GerÃ§ek Ã¶lÃ§ekte bu Ã§ok daha uzun sÃ¼rebilir.

Bir kez daha Ã§Ä±ktÄ± dosyalarÄ±nÄ± Ã§alÄ±ÅŸma dizininde bulabilirsiniz:

```bash
ls ENCSR000COQ1_1*
```

```console title="Ã‡Ä±ktÄ±"
ENCSR000COQ1_1_trimmed.bam  ENCSR000COQ1_1_trimmed.hisat2.log
```

### 2.5. Ã‡Ä±ktÄ± dosyalarÄ±nÄ± konteyner dÄ±ÅŸÄ±ndaki dosya sistemine taÅŸÄ±yÄ±n

```bash
mkdir /data/aligned
mv ENCSR000COQ1_1* /data/aligned
```

### 2.6. Konteynerden Ã§Ä±kÄ±n

```bash
exit
```

---

## 3. KapsamlÄ± bir QC raporu oluÅŸturun

`multiqc`'nin yÃ¼klÃ¼ olduÄŸu bir konteyner imajÄ±nÄ± Ã§ekeceÄŸiz, etkileÅŸimli olarak baÅŸlatacaÄŸÄ±z ve Ã¶ncesi/sonrasÄ± FastQC rapor dosyalarÄ± Ã¼zerinde bir rapor oluÅŸturma komutu Ã§alÄ±ÅŸtÄ±racaÄŸÄ±z.

### 3.1. `multiqc` konteynerini Ã§ekin

```bash
docker pull community.wave.seqera.io/library/pip_multiqc:a3c26f6199d64b7c
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console
    ad8f247edb55897c: Pulling from library/pip_multiqc
    dafa2b0c44d2: Already exists
    dec6b097362e: Already exists
    f88da01cff0b: Already exists
    4f4fb700ef54: Already exists
    92dc97a3ef36: Already exists
    403f74b0f85e: Already exists
    10b8c00c10a5: Already exists
    17dc7ea432cc: Already exists
    bb36d6c3110d: Already exists
    0ea1a16bbe82: Already exists
    030a47592a0a: Already exists
    3f229294c69a: Pull complete
    5a5ad47fd84c: Pull complete
    Digest: sha256:0ebb1d9605395a7df49ad0eb366b21f46afd96a5090376b0d8941cf5294a895a
    Status: Downloaded newer image for community.wave.seqera.io/library/pip_multiqc:a3c26f6199d64b7c
    community.wave.seqera.io/library/pip_multiqc:a3c26f6199d64b7c
    ```

### 3.2. `multiqc` konteynerini etkileÅŸimli olarak baÅŸlatÄ±n

```bash
docker run -it -v ./data:/data community.wave.seqera.io/library/pip_multiqc:a3c26f6199d64b7c
```

### 3.3. `multiqc` komutunu Ã§alÄ±ÅŸtÄ±rÄ±n

```bash
multiqc /data/reads /data/trimmed /data/aligned -n ENCSR000COQ1_1_QC
```

??? success "Komut Ã§Ä±ktÄ±sÄ±"

    ```console

    /// MultiQC ğŸ” v1.27.1

          file_search | Search path: /data/reads
          file_search | Search path: /data/trimmed
          file_search | Search path: /data/aligned
            searching | â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â” 100% 20/20
                hisat2 | Found 1 reports
              cutadapt | Found 1 reports
                fastqc | Found 1 reports
        write_results | Data        : ENCSR000COQ1_1_QC_data
        write_results | Report      : ENCSR000COQ1_1_QC.html
              multiqc | MultiQC complete
    ```

MultiQC, uyumlu QC raporlarÄ± iÃ§in dizinlerde arama yapabilir ve bulduÄŸu her ÅŸeyi birleÅŸtirir.

Burada aracÄ±n oluÅŸturduÄŸumuz Ã¼Ã§ QC raporunu da bulduÄŸunu gÃ¶rÃ¼yoruz: `fastqc` ile yaptÄ±ÄŸÄ±mÄ±z ilk QC, `cutadapt`'ten (via `trim_galore`) kÄ±rpma sonrasÄ± rapor ve `hisat2` tarafÄ±ndan Ã¼retilen hizalama sonrasÄ± QC.

Ã‡Ä±ktÄ± dosyalarÄ± bir kez daha Ã§alÄ±ÅŸma dizinindedir:

```bash
ls ENCSR000COQ1_1_QC*
```

```console title="Ã‡Ä±ktÄ±"
ENCSR000COQ1_1_QC.html

ENCSR000COQ1_1_QC_data:
cutadapt_filtered_reads_plot.txt                     fastqc_top_overrepresented_sequences_table.txt
cutadapt_trimmed_sequences_plot_3_Counts.txt         hisat2_se_plot.txt
cutadapt_trimmed_sequences_plot_3_Obs_Exp.txt        multiqc.log
fastqc-status-check-heatmap.txt                      multiqc_citations.txt
fastqc_adapter_content_plot.txt                      multiqc_cutadapt.txt
fastqc_per_base_n_content_plot.txt                   multiqc_data.json
fastqc_per_base_sequence_quality_plot.txt            multiqc_fastqc.txt
fastqc_per_sequence_gc_content_plot_Counts.txt       multiqc_general_stats.txt
fastqc_per_sequence_gc_content_plot_Percentages.txt  multiqc_hisat2.txt
fastqc_per_sequence_quality_scores_plot.txt          multiqc_software_versions.txt
fastqc_sequence_counts_plot.txt                      multiqc_sources.txt
fastqc_sequence_duplication_levels_plot.txt
```

### 3.4. Ã‡Ä±ktÄ± dosyalarÄ±nÄ± konteyner dÄ±ÅŸÄ±ndaki dosya sistemine taÅŸÄ±yÄ±n

```bash
mkdir /data/final_qc
mv ENCSR000COQ1_1_QC** /data/final_qc
```

### 3.5. Konteynerden Ã§Ä±kÄ±n

```bash
exit
```

---

### Ã‡Ä±karÄ±mlar

TÃ¼m bireysel komutlarÄ± ilgili konteynerlerde etkileÅŸimli olarak test ettiniz.

### SÄ±rada ne var?

AynÄ± komutlarÄ±, Ã§alÄ±ÅŸmayÄ± yÃ¼rÃ¼tmek iÃ§in konteynerler kullanan Ã§ok adÄ±mlÄ± bir iÅŸ akÄ±ÅŸÄ±na nasÄ±l sarmalayacaÄŸÄ±nÄ±zÄ± Ã¶ÄŸrenin.
